### ISSUE TITLE
calls cmd.exe directly without checking that it is in PATH

### ISSUE THREAD

**Issue by cslycord** [kou1okada/apt-cyg#42]
If /cygdrive/c/Windows/System32 is not in PATH, the dist-upgrade functionality doesn't fully work.

In the get_codepage function, the script directly calls cmd.exe without making sure that its directory is in PATH So the multiple calls to checking for codepage don't work.

After that, it's able to still run the setup.exe since but it hangs at the "ending cygwin install" part that is run inside a cmd DOS window.

**Comment by cslycord** [comment]
As well, I can confirm that running
`PATH=/cygdrive/c/Windows/System32:${PATH} apt-cyg dist-upgrade`

runs perfectly as expected.

**Comment by kou1okada** [comment]
Hmm,,,
I wonder why you drop `C:\Windows\System32` from PATH environment variable.
But It seems the problem which should be considered.

I want to know more details of conditions to fail dist-upgrade.
Can you teach me following results on your environment which fails `apt-cyg dist-upgrade`:
1. `cygstart cmd` on cygwin.
2. `echo $PATH` on cygwin.
3. `ECHO %PATH%` on `cmd.exe` which opend with `[Windows Key]+R cmd.exe`. 

I want to know if your PATH environment variable of Windows (not cygwin) includes 'C:\Windows\System32' or not.
Because on my environment, following situation on cygwin made no problems:
```
export PATH=/usr/bin # Drop `C:\Windows\System32` from PATH.
cygstart cmd
```
I also tried that remove every Path variables (for system and for user) with `System property dialog` on Windows.
But `cygstart cmd` still run correctly too.
Oh! it's amazing!!!

And unfortunately on my environment, following situation worked except only `chcp` did not work:
```
PATH=/usr/bin
~/local/bin/apt-cyg dist-upgrade
```
I must refer `$COMSPEC`, but `cygstart` seems to  have a strong robustness.

**Comment by kou1okada** [comment]
Oops!
```
>where chcp
C:\Windows\System32\chcp.com
```

**Comment by cslycord** [comment]
I had changed my cygwin PATH to just include standard UNIX paths. I had
done this because when I was compiling some software in cygwin, it was
catching windows executables earlier in the PATH and was giving errors
because it was making a mix of compiling for windows and cygwin. So, to
ensure that it wasn't using windows stuff, I changed the PATH in ~/.bashrc

And then I saw that apt-cyg would no longer update itself since it could no
longer find git (because it was previously using the windows git perfectly
fine), saw it could update itself when I installed the cygwin git package,
and decided to check if dist-upgrade would work, since it seemed to be the
only thing part I could find in your script that explicitly relied on
windows.

**Comment by kou1okada** [comment]
I want to know answers of 3 questions in my previous post.

**Comment by cslycord** [comment]
1) cygstart cmd opens up a dos shell but one with the %PATH% as I had set
in cygwin.
2) In cygwin, PATH=/home/user/ffmpeg/bin:/usr/local/bin:/usr/bin
Note: this isn't the same PATH that I had set on the system I reported this
from. But the behavior is the same.
3) C:\Program Files (x86)\Intel\Intel(R) Management Engine
Components\iCLS\;C:\Program Files\Intel\Intel(R) Management Engine
Components\iCLS\;C:\windows\system32;C:\windows;C:\windows\System32\Wbem;C:\windows\System32\WindowsPowerShell\v1.0\;C:\windows\System32\OpenSSH\;C:\Program
Files (x86)\Intel\Intel(R) Management Engine Components\DAL;C:\Program
Files\Intel\Intel(R) Management Engine
Components\DAL;C:\ProgramData\chocolatey\bin;C:\Users\user\AppData\Local\Microsoft\WindowsApps;C:\Program
Files\Bandizip\

**Comment by kou1okada** [comment]
Thanks.
As expected, `cygstart` worked well despite that PATH did not have `C:\Windows\System32`, didn't it.

I think that your `dist-upgrade` failed at calling `chcp`, maybe.
My last patch 3708975e9634fda3593e985f64413dcfba41e07a (it's on issue-42 branch) may solve your problem on `dist-upgrade`, but I can not be sure.
Can you try it?

If you still have the problem, would you please report it?
I will try to solve it.

**Comment by cslycord** [comment]
With that change, the upgrade process finishes okay.

The only issue is that after choosing "Y" to do the upgrade, inside the
cygwin terminal I see:
'DOSKEY' is not recognized as an internal or external command,
operable program or batch file.
'DOSKEY' is not recognized as an internal or external command,
operable program or batch file.

And the same error appears once in the top of the new DOS shell.

But, otherwise the upgrade process works

**Comment by kou1okada** [comment]
In your last report, did `cygstart cmd` not show these warnings of` DOSKEY`?

**Comment by cslycord** [comment]
Running "cygstart cmd" does result in that message appearing once in the
DOS shell before bringing up the shell prompt.

**Comment by kou1okada** [comment]
Maybe, the first two warning in cygwin terminal occured at [`findworkspace` funcion](https://github.com/kou1okada/apt-cyg/blob/3708975e9634fda3593e985f64413dcfba41e07a/apt-cyg#L273-L304) to call [`setuprc_get_section last-mirror`](https://github.com/kou1okada/apt-cyg/blob/3708975e9634fda3593e985f64413dcfba41e07a/apt-cyg#L273-L304) and [`setuprc_get_section last-cache`](https://github.com/kou1okada/apt-cyg/blob/3708975e9634fda3593e985f64413dcfba41e07a/apt-cyg#L276).
[`setuprc_get_section`](Probably, my COMSPEC does not call DOSKEY.) function calls [`cp2utf8`](https://github.com/kou1okada/apt-cyg/blob/3708975e9634fda3593e985f64413dcfba41e07a/apt-cyg#L1159-L1162) and it call [`"${COMSPEC//\\//}" /c "$SYSTEMROOT"'\System32\chcp.com'`](https://github.com/kou1okada/apt-cyg/blob/3708975e9634fda3593e985f64413dcfba41e07a/apt-cyg#L1156).
And maybe the last one in DOS Prompt occured at `dist-upgrade` function to call [`cygstart "$COMSPEC"`](https://github.com/kou1okada/apt-cyg/blob/3708975e9634fda3593e985f64413dcfba41e07a/apt-cyg#L1272).

Total three `$COMSPEC` failed to call `DOSKEY`, maybe.

However, I tried to run `dist-upgrade` after dropping all the paths on the Windows system, but I could not see these warnings of `DOSKEY`.
Probably, my `$COMSPEC` does not call `DOSKEY`.

Perhaps you are using `AUTOEXEC.BAT` and you are initializing `DOSKEY` in it?

**Comment by kou1okada** [comment]
Hmm, it seems that not `AUTOEXEC.BAT` but `Autorun` registory entries.
https://superuser.com/questions/1134368/create-permanent-doskey-in-windows-cmd

Run following commands and report result is help me to fix this issue:
```
readarray -t entries < <(find /proc/registry/{HKEY_LOCAL_MACHINE,HKEY_CURRENT_USER}/SOFTWARE/Microsoft/Command\ Processor/ -iname 'autorun')
if (( ${#entries} <= 0 )); then
  echo "entries not found."
else
  for i in "${entries[@]}"; do
    echo "----------"
    echo "$i"
    ls -l "$i"
    cat "$i"
    od -tx1a "$i"
  done
fi
```

But the `%PATH%` should be resolved by `Autorun` if it is required by `Autorun`.

**Comment by cslycord** [comment]
You hit the nail on the head. My Autorun is calling a batch script that
uses DOSKEY.

**Comment by kou1okada** [comment]
Added a commit for improvement b7acbae35c27071d9fbbe363823c4ef58c4035d0 to issue-42, and merged it to master  at c518bc83f86270eea4f3ad3848b8e2b354284420.
